arch/apple2/encoder.cc

```diff
@@ -0,0 +1,187 @@
+#include "globals.h"
+#include "arch/apple2/apple2.h"
+#include "decoders/decoders.h"
+#include "encoders/encoders.h"
+#include "sector.h"
+#include "writer.h"
+#include "image.h"
+#include "fmt/format.h"
+#include "lib/encoders/encoders.pb.h"
+#include <ctype.h>
+#include "bytes.h"
+
+static int encode_data_gcr(uint8_t data) {
+    switch(data)
+    {
+	#define GCR_ENTRY(gcr, data) \
+	case data: return gcr;
+	#include "data_gcr.h"
+	#undef GCR_ENTRY
+    }
+    return -1;
+}
+
+class Apple2Encoder : public AbstractEncoder
+{
+public:
+    Apple2Encoder(const EncoderProto& config):
+	AbstractEncoder(config)
+    {}
+
+public:
+    std::vector<std::shared_ptr<const Sector>> collectSectors(int physicalTrack, int physicalSide, const Image& image) override
+    {
+	std::vector<std::shared_ptr<const Sector>> sectors;
+	constexpr auto numSectors = 16;
+	if (physicalSide == 0)
+	{
+	    int logicalTrack = physicalTrack / 2;
+	    unsigned numSectors = 16;
+	    for (int sectorId=0; sectorId<numSectors; sectorId++)
+	    {
+		const auto& sector = image.get(logicalTrack, 0, sectorId);
+		if (sector)
+		    sectors.push_back(sector);
+	    }
+	}
+
+	return sectors;
+    }
+
+    std::unique_ptr<Fluxmap> encode(int physicalTrack, int physicalSide,
+	    const std::vector<std::shared_ptr<const Sector>>& sectors, const Image& image) override
+    {
+	if (physicalSide != 0)
+	    return std::unique_ptr<Fluxmap>();
+
+	int logicalTrack = physicalTrack / 2;
+	double clockRateUs = 4.;
+
+	int bitsPerRevolution = 200000.0 / clockRateUs;
+
+	std::vector<bool> bits(bitsPerRevolution);
+	unsigned cursor = 0;
+
+	for (const auto& sector : sectors) {
+	    if(sector) {
+		writeSector(bits, cursor, *sector);
+	    }
+	}
+
+	if (cursor >= bits.size())
+	    Error() << fmt::format("track data overrun by {} bits", cursor - bits.size());
+	fillBitmapTo(bits, cursor, bits.size(), { true, false });
+
+	std::unique_ptr<Fluxmap> fluxmap(new Fluxmap);
+	fluxmap->appendBits(bits, clockRateUs*1e3);
+	return fluxmap;
+
+    }
+
+private:
+    uint8_t volume_id = 254;
```

Comment:
Is this ever likely to need to change? Is it worth putting it in the proto file as a parameter?

Reply:
It may need to be. But, can we leave it to the future?

I have a limited selection of disks to check, but DOS 3.3 and a disk created from it with `INIT` both come up VOLUME 254. However, ProDOS comes up as VOLUME 1 (when listed from inside DOS 3.3 `CATALOG` which shows it; ProDOS doesn't seem to show the volume number).

For DOS 3.3 the volume number can be detected in the VTOC track 17 sector 0 byte 6. It can also be specified to INIT:
```
Each DOS 3.3 diskette has a Volume Number set at the time it is initialized. The
default Volume Number used by INIT is 254. You can set the number to something
else at the time a diskette is initialized:

INIT HELLO, V19```
for prodos I didn't find how to format with a different volume ID than 1, though a disk converted from DOS 3.3 _to_ ProDOS may retain its original volume ID.

Reply:
Yes, no problem.